File: packages/react-native-gtkx/tests/gtk/support/virtual-stylus.py
#!/usr/bin/env python3
"""A virtual pressure-sensitive stylus/tablet on /dev/uinput, driven over stdin.

The other half of ./virtual-stylus.ts, which is the file to read first, and a
sibling of ./virtual-touchpad.py built the same way: the kernel is the
injection point, so libinput and the compositor above it classify the device as
real hardware and run their normal tablet code over it. Node has no ioctl,
hence Python; python3-evdev wraps UI_DEV_SETUP/UI_SET_EVBIT/... .

The device shape is a Wacom Intuos5 touch M pen tablet, copied from libinput's
own litest descriptor (litest-device-wacom-intuos5-pen.c) down to the USB ids.
Two things in that shape are load-bearing:

  * BTN_TOOL_PEN together with ABS_X/ABS_Y is precisely what systemd-udev's
    input_id builtin keys ID_INPUT_TABLET off, and ID_INPUT_TABLET is what
    makes libinput build a tablet dispatch instead of a pointer one.
  * INPUT_PROP_POINTER marks it an *external* tablet (an Intuos, not a Cintiq),
    which is what makes the compositor map the whole tablet area onto the whole
    screen. A device tagged INPUT_PROP_DIRECT would want an output to be bound
    to, and gets no sensible mapping on a VM with no real tablet configured.

Coordinates on stdin are fractions 0..1 of the tablet area, so 0.5 0.5 is the
middle of the screen; pressure is likewise a fraction 0..1 of the axis range.

    prox_in <x> <y>                          BTN_TOOL_PEN down, pressure 0
    move <x> <y> <p>                         one frame
    ramp <x0> <y0> <x1> <y1> <p0> <p1> <n>   n frames, pressure interpolated
    prox_out
    quit

Protocol otherwise matches ./virtual-touchpad.py exactly: one command per line
on stdin, one `ok` per command on stdout once the injected sequence has been
written and flushed, and one `ready` before any of it.
"""

import sys
import time

from evdev import AbsInfo, UInput
from evdev import ecodes as e

# Intuos5 touch M: 223.5mm x 139.7mm at 200 units/mm. libinput wants the
# resolution to convert to millimetres; without it, it warns and guesses.
RESOLUTION = 200
X_MAX = 44704
Y_MAX = 27940
# The absolute pressure range is arbitrary as far as the stack is concerned --
# libinput normalises pressure to 0.0..1.0 against this range, and the Wayland
# tablet protocol re-quantises to 0..65535 on the wire -- but a wide range
# makes a lost low bit obvious rather than invisible.
PRESSURE_MAX = 65535
DISTANCE_MAX = 63
TILT_MIN, TILT_MAX = -64, 63

CAPABILITIES = {
    e.EV_KEY: [
        e.BTN_TOOL_PEN,
        e.BTN_TOOL_RUBBER,
        e.BTN_TOOL_BRUSH,
        e.BTN_TOOL_PENCIL,
        e.BTN_TOOL_AIRBRUSH,
        e.BTN_TOUCH,
        e.BTN_STYLUS,
        e.BTN_STYLUS2,
    ],
    e.EV_ABS: [
        (e.ABS_X, AbsInfo(0, 0, X_MAX, 4, 0, RESOLUTION)),
        (e.ABS_Y, AbsInfo(0, 0, Y_MAX, 4, 0, RESOLUTION)),
        (e.ABS_PRESSURE, AbsInfo(0, 0, PRESSURE_MAX, 0, 0, 0)),
        (e.ABS_DISTANCE, AbsInfo(0, 0, DISTANCE_MAX, 0, 0, 0)),
        (e.ABS_TILT_X, AbsInfo(0, TILT_MIN, TILT_MAX, 0, 0, 57)),
        (e.ABS_TILT_Y, AbsInfo(0, TILT_MIN, TILT_MAX, 0, 0, 57)),
        (e.ABS_MISC, AbsInfo(0, 0, 0, 0, 0, 0)),
    ],
    # MSC_SERIAL is how a tablet identifies *which* pen is on it. libinput
    # turns it into the tool serial, and the compositor into a GdkDeviceTool.
    e.EV_MSC: [e.MSC_SERIAL],
}

FRAME_MS = 12
# Any non-zero serial; 0 means "no serial reported" to libinput.
TOOL_SERIAL = 0x00BEEF01
# ABS_MISC carries the Wacom tool id. 0x0802 is the Intuos Grip Pen, which is
# what litest reports; libinput maps it to LIBINPUT_TABLET_TOOL_TYPE_PEN.
TOOL_ID = 0x0802


class Stylus:
    def __init__(self):
        self.ui = UInput(
            CAPABILITIES,
            name="rn-gtkx virtual stylus",
            # Wacom's vendor id with the Intuos5 touch M product id. Claiming a
            # tablet libwacom already knows means libinput gets a real entry
            # out of its database instead of logging "device is not known".
            vendor=0x056A,
            product=0x0027,
            version=0x0110,
            phys="rn-gtkx/virtual-stylus/input0",
            input_props=[e.INPUT_PROP_POINTER],
        )
        self.in_proximity = False

    def sync(self):
        self.ui.syn()
        time.sleep(FRAME_MS / 1000)

    @staticmethod
    def _abs(frac, maximum):
        return max(0, min(maximum, int(round(frac * maximum))))

    def prox_in(self, x, y):
        self.ui.write(e.EV_ABS, e.ABS_X, self._abs(x, X_MAX))
        self.ui.write(e.EV_ABS, e.ABS_Y, self._abs(y, Y_MAX))
        self.ui.write(e.EV_ABS, e.ABS_DISTANCE, 20)
        # Proximity must be entered at zero pressure. libinput watches the
        # first in-proximity frames for a non-zero resting pressure and, if it
        # sees one, decides the pen has a miscalibrated tip and silently
        # subtracts that value as an offset from everything that follows.
        self.ui.write(e.EV_ABS, e.ABS_PRESSURE, 0)
        self.ui.write(e.EV_ABS, e.ABS_TILT_X, 0)
        self.ui.write(e.EV_ABS, e.ABS_TILT_Y, 0)
        self.ui.write(e.EV_ABS, e.ABS_MISC, TOOL_ID)
        self.ui.write(e.EV_MSC, e.MSC_SERIAL, TOOL_SERIAL)
        self.ui.write(e.EV_KEY, e.BTN_TOOL_PEN, 1)
        self.sync()
        self.in_proximity = True

    def move(self, x, y, pressure):
        value = self._abs(pressure, PRESSURE_MAX)
        self.ui.write(e.EV_ABS, e.ABS_X, self._abs(x, X_MAX))
        self.ui.write(e.EV_ABS, e.ABS_Y, self._abs(y, Y_MAX))
        self.ui.write(e.EV_ABS, e.ABS_PRESSURE, value)
        # Distance and pressure are mutually exclusive on a real tablet: the
        # pen either hovers at a distance or rests with a pressure.
        self.ui.write(e.EV_ABS, e.ABS_DISTANCE, 0 if value else 20)
        # BTN_TOUCH is the tip switch, and it is what GtkGestureStylus turns
        # into `down`/`up` -- so a zero-pressure frame is a hover, and the
        # first non-zero one is the press.
        self.ui.write(e.EV_KEY, e.BTN_TOUCH, 1 if value else 0)
        self.ui.write(e.EV_MSC, e.MSC_SERIAL, TOOL_SERIAL)
        self.sync()

    def ramp(self, x0, y0, x1, y1, p0, p1, steps):
        for i in range(steps + 1):
            t = i / steps
            self.move(x0 + (x1 - x0) * t, y0 + (y1 - y0) * t, p0 + (p1 - p0) * t)

    def prox_out(self):
        self.ui.write(e.EV_ABS, e.ABS_PRESSURE, 0)
        self.ui.write(e.EV_KEY, e.BTN_TOUCH, 0)
        self.sync()
        self.ui.write(e.EV_ABS, e.ABS_DISTANCE, 0)
        self.ui.write(e.EV_ABS, e.ABS_MISC, 0)
        self.ui.write(e.EV_MSC, e.MSC_SERIAL, TOOL_SERIAL)
        self.ui.write(e.EV_KEY, e.BTN_TOOL_PEN, 0)
        self.sync()
        self.in_proximity = False


def main():
    pen = Stylus()
    # udev has to see the device and the compositor's libinput context has to
    # add it to the seat before anything injected through it means anything.
    # The caller waits again on top of this and burns a throwaway proximity
    # cycle before it measures anything -- see ./virtual-stylus.ts, which
    # explains why one settle is not enough for a tablet.
    time.sleep(1.5)
    print("ready", flush=True)
    for line in sys.stdin:
        parts = line.split()
        if not parts or parts[0] == "quit":
            break
        command, args = parts[0], [float(a) for a in parts[1:]]
        try:
            if command == "prox_in":
                pen.prox_in(args[0], args[1])
            elif command == "move":
                pen.move(args[0], args[1], args[2])
            elif command == "ramp":
                pen.ramp(*args[:6], int(args[6]))
            elif command == "prox_out":
                pen.prox_out()
            else:
                print(f"error unknown command {command}", flush=True)
                continue
        except Exception as error:  # noqa: BLE001 - reported, not swallowed
            print(f"error {error}", flush=True)
            continue
        print("ok", flush=True)
    # A pen left in proximity outlives this process in the compositor's idea of
    # the seat, so the exit path puts it down whatever happened above.
    if pen.in_proximity:
        pen.prox_out()
    pen.ui.close()


if __name__ == "__main__":
    main()

File: packages/react-native-gtkx/tests/gtk/support/virtual-stylus.ts
// Real tablet input, for the one axis no pointer and no touchpad can carry.
//
// Why this exists, and why it is not `virtual-pointer.ts` with an extra
// argument: `wl_pointer` has no pressure axis and never will. Pressure lives
// in a different Wayland protocol entirely (`zwp_tablet_v2`), and a client
// cannot synthesize into it — there is no `zwlr_virtual_tablet_v1` in wlroots
// or anywhere else. Pressure is something libinput READS off a device it has
// classified as a tablet, exactly as a pinch is something it CONCLUDES from
// two fingers on a device it has classified as a touchpad.
//
// So the injection point is the same one `./virtual-touchpad.ts` uses, one
// layer below Wayland: a virtual pen tablet on `/dev/uinput`. The chain is
// then entirely real:
//
//   uinput -> kernel evdev -> libinput (TABLET_TOOL_AXIS/TIP/PROXIMITY)
//          -> compositor (zwp_tablet_tool_v2.pressure, 0..65535)
//          -> GDK (GdkDeviceTool, GDK_AXIS_PRESSURE normalised to [0, 1])
//          -> GtkGestureStylus.get_axis(Gdk.AxisUse.PRESSURE)
//
// THREE THINGS WERE MEASURED ON THIS CHAIN, and a caller has to know all
// three, because each one changes what an assertion may say:
//
// 1. It works under a compositor with a libinput backend and delivers nothing
//    at all under one without — the same split `./virtual-touchpad.ts`
//    documents, for the same reason. The headless sway `@gtkx/vitest` starts
//    per worker is launched with `WLR_BACKENDS=headless` and
//    `WLR_LIBINPUT_NO_DEVICES=1`, so it enumerates zero input devices and a
//    uinput tablet is invisible to it. Under the VM's real GNOME session the
//    same injection reaches `GtkGestureStylus` with varying pressure.
//    `spike/gesture-detector/run-stylus.sh` is what runs against the real one.
//
// 2. THE CURVE IS NOT LINEAR. mutter applies a transfer curve between the
//    libinput reading and the Wayland wire, and it measured as roughly
//    QUADRATIC: the pressure GTK reports is about the SQUARE of the fraction
//    injected here (inject 0.5, GTK says ~0.25; inject 1.0, GTK says ~1.0).
//    So a test may assert MONOTONICITY and the ENDPOINTS and nothing in
//    between — asserting that GTK sees the number that went in would be
//    asserting that this particular compositor has no curve, which is not a
//    property of the platform.
//
// 3. **THE DEVICE HAS TO EXIST BEFORE THE CLIENT DOES**, and this is the
//    constraint that shapes every caller. It was measured on the wire
//    (`WAYLAND_DEBUG=1` on the client) against `libinput debug-events` one
//    layer below, and libinput is never the problem: it emitted every injected
//    cycle, complete, correct and linear, in every run that follows. What
//    varies is whether mutter forwards one. Same client, same device, same
//    injected ramp — only the startup order changed:
//
//      device, throwaway cycle, then client -> 24 samples, 0.0016 .. 1.0000
//      client, then device                  -> 0 samples
//      client, then device, throwaway cycle -> 0 samples
//
//    A client that is already connected when the tablet appears is told
//    `tablet_added` and then never told `tool_added`, so no tool ever enters
//    proximity for it. A client that binds `zwp_tablet_seat_v2` after the
//    device exists is told both. A probe that is itself the GTK client
//    therefore has to open the device BEFORE it starts its application —
//    `spike/gesture-detector/src/probe-stylus.tsx` does exactly that and says
//    so at its entry point.
//
//    Two smaller findings sit under that one. The throwaway cycle is still
//    needed, because the first proximity after a hotplug is swallowed even in
//    the working order; `createVirtualStylus` burns it, which is safe
//    precisely because there is no client yet. And a cycle that starts within
//    about a second of the previous `proximity_out` is dropped in full —
//    `proximity_in`, `down`, every axis event, `up` — so the rule for callers
//    is ONE `proximityIn`, everything inside it, ONE `proximityOut`. The pen
//    may be pressed and lifted as often as a test likes without leaving
//    proximity, which is how several zones get measured in one run and is what
//    a real pen does anyway.
//
// COORDINATES ARE SCREEN FRACTIONS, not widget pixels, and that follows from
// the device shape: `INPUT_PROP_POINTER` marks an EXTERNAL tablet, which the
// compositor maps whole-area-onto-whole-screen. `0.5, 0.5` is the middle of
// the display. Only a FULLSCREEN surface is guaranteed to be under every
// coordinate a caller can name, so a test that aims at a widget should
// fullscreen its window first and work out fractions from the widget's rect
// over the toplevel's size.
//
// The device itself is created from Python, in `./virtual-stylus.py`: uinput
// needs ioctls and Node has none. This half is the process and the line
// protocol, in the shape `createVirtualPointer` and `createVirtualTouchpad`
// already established, so a test drives a pen the same way it drives the
// other two.
import {
  spawn,
  spawnSync,
  type ChildProcessWithoutNullStreams,
} from "node:child_process"
import { existsSync } from "node:fs"
import { dirname, join } from "node:path"
import { createInterface, type Interface } from "node:readline"

/** Where the helper sits in the repo, from the repo root. */
const HELPER_FROM_ROOT = join(
  "packages",
  "react-native-gtkx",
  "tests",
  "gtk",
  "support",
  "virtual-stylus.py",
)

/**
 * The sibling `.py`, or the repo copy of it.
 *
 * A test running from source finds it next to this file, which is the whole of
 * it. A BUNDLED consumer does not: `spike/gesture-detector` builds this module
 * into `dist/bundle.mjs`, where `import.meta.dirname` is the bundle's directory
 * and there is no sibling to find — the helper is a Python file, so no bundler
 * will ever carry it along. Walking up for the repo path covers that without
 * an environment variable nobody would remember to set.
 */
const resolveHelper = (): string | null => {
  const sibling = join(import.meta.dirname, "virtual-stylus.py")
  if (existsSync(sibling)) {
    return sibling
  }
  for (const start of [import.meta.dirname, process.cwd()]) {
    let directory = start
    for (;;) {
      const candidate = join(directory, HELPER_FROM_ROOT)
      if (existsSync(candidate)) {
        return candidate
      }
      const parent = dirname(directory)
      if (parent === directory) {
        break
      }
      directory = parent
    }
  }
  return null
}

const HELPER = resolveHelper()

/** How long a single injected sequence may take before it is given up on. */
const COMMAND_TIMEOUT_MS = 15_000

/**
 * How long to wait, on top of the helper's own udev settle, before the tablet
 * is usable — the window mutter needs to finish adding it to the seat. Both
 * halves together are the 4.5s the working rig this was ported from waits.
 */
const TABLET_SETTLE_MS = 3000

/**
 * How long to wait after the throwaway cycle's `proximity_out` before handing
 * the pen over. Matches the `sleep 1` the reference rig has in the same place;
 * a caller that then starts a GTK application adds several seconds more.
 */
const PROXIMITY_COOLDOWN_MS = 1000

export class VirtualStylusUnavailable extends Error {}

/** One frame's worth of pen state. Fractions of the screen, and of `[0, 1]`. */
export type StylusPoint = {
  x: number
  y: number
  /**
   * The fraction of the pressure axis to INJECT, which is not the number GTK
   * will report — see point 2 in the header.
   */
  force: number
}

export type VirtualStylus = {
  /**
   * Brings the pen into hover range over `(x, y)` with the tip UP.
   *
   * Separate from touching because the device makes it separate and because
   * the distinction is the point: a pen in proximity produces motion with no
   * pressure, which is what `GtkGestureStylus` reports as `proximity`/`motion`
   * and never as `down`.
   */
  proximityIn(x: number, y: number): Promise<void>
  /**
   * One frame with the pen at `(x, y)` under `force`.
   *
   * A `force` of 0 lifts the tip (BTN_TOUCH goes low) and is how a press is
   * ENDED — that is the `up` `GtkGestureStylus` reports.
   */
  moveTo(x: number, y: number, force: number): Promise<void>
  /**
   * `steps` frames interpolating position and pressure linearly from `from` to
   * `to`, which is the only way to produce a pressure RAMP: one frame per
   * value, at the device's frame interval, so libinput sees a plausible pen
   * rather than a jump it would filter.
   */
  ramp(from: StylusPoint, to: StylusPoint, steps?: number): Promise<void>
  /** Takes the pen out of range. Ends the tablet tool's proximity. */
  proximityOut(): Promise<void>
  dispose(): void
}

/**
 * Everything that has to be true before a virtual stylus can exist, checked up
 * front so a test can skip on an environment difference rather than fail on
 * one. Returns the reason it cannot, or null.
 *
 * Deliberately NOT checked here: whether the compositor has a libinput
 * backend. There is no way to ask one, and the failure it produces is a
 * silence rather than an error — which is why every caller of this needs a
 * negative control and an explicit runner, not a probe for the backend.
 */
export const virtualStylusBlocker = (): string | null => {
  if (process.platform !== "linux") {
    return "not Linux"
  }
  if (!existsSync("/dev/uinput")) {
    return "no /dev/uinput — the kernel module is not loaded"
  }
  if (HELPER === null) {
    return `the uinput helper is missing (${HELPER_FROM_ROOT})`
  }
  // Root, because /dev/uinput is root:input 0660 and the test user is not in
  // the `input` group on any machine this has run on.
  if (spawnSync("sudo", ["-n", "true"]).status !== 0) {
    return "no passwordless sudo, and /dev/uinput is root:input 0660"
  }
  if (spawnSync("python3", ["-c", "import evdev"]).status !== 0) {
    return "python3-evdev is not importable"
  }
  return null
}

type Pending = {
  resolve: () => void
  reject: (error: Error) => void
  timer: ReturnType<typeof setTimeout>
}

const settle = (pending: Pending[], line: string): void => {
  const next = pending.shift()
  if (!next) {
    return
  }
  clearTimeout(next.timer)
  if (line.startsWith("error")) {
    next.reject(new Error(`virtual stylus: ${line}`))
    return
  }
  next.resolve()
}

/**
 * Creates the tablet, waits for the compositor to have added it, and burns the
 * proximity cycle mutter swallows.
 *
 * Throws {@link VirtualStylusUnavailable} when the machine cannot host one, so
 * a test can skip rather than fail on an environment difference — the same
 * contract `createVirtualTouchpad` and `createVirtualPointer` have.
 */
export const createVirtualStylus = async (): Promise<VirtualStylus> => {
  const blocker = virtualStylusBlocker()
  if (blocker !== null) {
    throw new VirtualStylusUnavailable(blocker)
  }

  const child: ChildProcessWithoutNullStreams = spawn(
    "sudo",
    // `blocker` is null, so the helper was found.
    ["-n", "python3", HELPER as string],
    { stdio: ["pipe", "pipe", "pipe"] },
  )
  const pending: Pending[] = []
  let stderr = ""
  child.stderr.setEncoding("utf8")
  child.stderr.on("data", (chunk: string) => {
    stderr += chunk
  })

  const lines: Interface = createInterface({ input: child.stdout })
  const ready = new Promise<void>((resolve, reject) => {
    const onExit = (): void => {
      reject(
        new VirtualStylusUnavailable(
          `the uinput helper exited before it was ready: ${stderr.trim()}`,
        ),
      )
    }
    child.once("exit", onExit)
    lines.once("line", (line: string) => {
      child.off("exit", onExit)
      if (line.trim() === "ready") {
        resolve()
        return
      }
      reject(new VirtualStylusUnavailable(`the uinput helper said: ${line}`))
    })
  })
  await ready

  lines.on("line", (line: string) => {
    settle(pending, line.trim())
  })
  child.on("exit", () => {
    while (pending.length > 0) {
      settle(pending, "error the uinput helper exited")
    }
  })

  const send = (command: string): Promise<void> =>
    new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error(`virtual stylus: \`${command}\` did not finish`))
      }, COMMAND_TIMEOUT_MS)
      pending.push({ resolve, reject, timer })
      child.stdin.write(`${command}\n`)
    })

  const stylus: VirtualStylus = {
    proximityIn: (x, y) => send(`prox_in ${x} ${y}`),
    moveTo: (x, y, force) => send(`move ${x} ${y} ${force}`),
    ramp: (from, to, steps = 20) =>
      send(
        `ramp ${from.x} ${from.y} ${to.x} ${to.y} ${from.force} ${to.force} ${steps}`,
      ),
    proximityOut: () => send("prox_out"),
    dispose() {
      for (const entry of pending) {
        clearTimeout(entry.timer)
      }
      pending.length = 0
      child.stdin.end("quit\n")
      lines.close()
      // The helper closes the uinput device on its way out, which is what
      // makes the kernel remove it. If it will not go, the device would
      // outlive the test run.
      const killer = setTimeout(() => {
        child.kill("SIGKILL")
      }, 2000)
      child.once("exit", () => {
        clearTimeout(killer)
      })
    },
  }

  // Point 3 in the header, and the difference between a run that measures
  // something and a run that measures silence: settle, spend one whole
  // proximity cycle on nothing, and then wait again before handing the pen
  // over. Both waits were measured; neither is a guess.
  await new Promise<void>((resolve) => {
    setTimeout(resolve, TABLET_SETTLE_MS)
  })
  await stylus.proximityIn(0.5, 0.5)
  await stylus.moveTo(0.5, 0.5, 0.3)
  await stylus.moveTo(0.5, 0.5, 0)
  await stylus.proximityOut()
  await new Promise<void>((resolve) => {
    setTimeout(resolve, PROXIMITY_COOLDOWN_MS)
  })

  return stylus
}
